### ISSUE TITLE
Velocity Extension

### ISSUE THREAD

**Issue by HYY-ying** [RaDelft/RaDelft-Dataset#8]
![img_v3_02fu_66eb061d-cdba-4329-a116-e8925d34d79g](https://github.com/user-attachments/assets/8be592e4-f9ea-401a-a86e-832e038749df)
![image](https://github.com/user-attachments/assets/34194fe8-b301-440a-be64-f7e226e29906)
Hello! I want to use the maximum Doppler unamiguous algorithm, how can I get the input "cube" for function cube_to_pointcloud?  I need velocity information of radar pointclouds. Is it incorrect to obtain Doppler velocity information directly from the cfar algorithm in the file generatePointCloud.m?

**Comment by HYY-ying** [comment]
![image](https://github.com/user-attachments/assets/1d7d5e0c-7ada-407d-b0da-f9ac862f1035)
![image](https://github.com/user-attachments/assets/690e75d2-15cb-45e7-827e-886b6684a744)
I also want to know how did you get the Maximum Unambiguous Velocity (without extension) . Why is it 2.93m/s? The maximumVelocity calculated in the file params.m is 29.6765m/s. The relationship between the two speeds is not a factor of 12, which is the number of all transmitters. If the TDM mechanism has an effect on the maximum unambiguous speed, the relationship between the two speeds should be a multiple of 12, I think.

**Comment by IgnacioRoldan** [comment]
Regarding the first question, you can get the cubes doing what is written in the screenshot you passed: run the generateCubesFFTs.m with the parameter saveDopplerFold equal to 1. If you use the provided point clouds, they do not have the velocity extension algorithm applied.

Regarding the maximum velocity, the algorithm I use to extend it (based on references [46 and 47]), is based on the overlapping virtual elements. This extends the maximum unambiguous velocity by a factor of 7.

**Comment by HYY-ying** [comment]
![image](https://github.com/user-attachments/assets/dd78d3e6-bb4f-4f5a-97aa-5ab1a951f440)
Thanks for your reply! My aim is to obtain radar point clouds with velocity through traditional signal processing algorithms, while the maximum unambiguous velocity needs to be extended. I'm not trying to replicate the deep network framework. I see your speed expansion algorithm in the function cube_to_pointcloud from data_preparation.py. How should I use this function to achieve speed expansion？How to get the cube and radarcube parameters in the red box respectively？

**Comment by IgnacioRoldan** [comment]
The deep network framework and the velocity extension algorithm are unrelated. 

To use the cube_to_pointcloud function with the unambiguous speed extension, you must first run the algorithm to unfold the Doppler. This is done using the generateCubesFFTs.m script with the parameter saveDopplerFold equal to 1. When you run it, it will generate a matrix with the corresponding Doppler fold per range cell. 

After you have these new files, you will have to call cube_to_pointcloud with the parameter dop_fold_path pointing to the newly generated files.

**Comment by HYY-ying** [comment]
Yeah, I've got the unfolded doppler data. So what data should I pass to the "cube" parameter and what data should I pass to the "radar_cube" parameter for the function cube_to_pointcloud?

**Comment by IgnacioRoldan** [comment]
That function takes a cube (or mask) as the parameter 'cube' (with the probability of detection if it comes from the network or binary if it is a lidar cube). If it is a lidar cube, the other parameters are not needed. If it is a radar cube, it needs the original cube (the one with powers, not binary) to take the power from. Then, also can take the file to the Doppler matrices to get the extended velocity.

However, I think I understand now what you want. You want to generate the point clouds with a CFAR algorithm but also use the Doppler extension. In that case, you have to code it yourself, but it shouldn't be too difficult. You can modify the _generatePointcloud.m_ file to read the Doppler matrix and find the corresponding speed, similar to lines 416-419 in the _data_preparation.py_ (in the function you mentioned)

**Comment by HYY-ying** [comment]
Yes! What I want is what you said, so I will try to change the code myself first. I am very interested in your data set, but there are many things I don't understand in the process of using it. Thank you very much for your patient answer！

**Comment by HYY-ying** [comment]
![image](https://github.com/user-attachments/assets/a4f2998c-f537-42ec-ad1b-bfd44ed6c8ab)
![f96acf2d4d6e6fe1dd632f22d0b608d7](https://github.com/user-attachments/assets/8b1e6826-b272-4a53-8101-e5ad6bc77359)
I still have a question, how did you get this maximum unambiguous velocity(without extention) 2.93m/s, was it experimentally measured or theoretically calculated? Using the formula for calculating the maximum unambiguous velocity, we can obtain that the maximum velocity before TDM is 29.6765m/s from your matlab code. There are 12 transmitters in total, and theoretically the maximum unambiguous speed should be one twelfth of the original speed after adopting the TDM mechanism, that is 29.6765 / 12≈2.47m/s in theory. Why is it actually 2.93m/s?

**Comment by IgnacioRoldan** [comment]
You are right; there is a mistake in the calculation. I will explain:

The maximum velocity (with TDMA) is $v_{max} = \frac{\lambda}{4T_c N_{tx}}$. In the paper, I used $T_c = 28\mu s$, but in reality, the total chirp duration is the ramp time + the idle time (the time that the system takes to go to the original frequency). The idle time is $5\mu s$ so $v_{max} = \frac{\lambda}{4T_c N_{tx}} = \frac{3.928e^{-3}}{4 * (28+5)e^{-6} * 12} = 2.48$  m/s. If you don't use the idle time (the extra $5\mu s$) is when you get the 2.93m/s.

In summary, the maximum speed should be 2.48m/s without extension and 17.36m/s with the extension.

**Comment by HYY-ying** [comment]
So the maximum speed of the radar data in the data set after using the speed expansion algorithm mentioned in the paper is actually only 17.36m/s, or 62.5km/h? The speed value of 20.56m/s in the paper is out of reach?

**Comment by IgnacioRoldan** [comment]
Yes, unfortunately, with the current extension method, that is the maximum unambiguous velocity. You can try to implement other methods. This will be corrected in the paper.
